File: numerical_methods/edo_ivp/leapfrog_O1/leap_frog.f90
! Programa elaborado por Jesús Eduardo Loera Casas
! Fecha 01/02/23

! En este programa se implementa el método de leap-frog o del salto de 
! rana para resolver una ecuación diferencial de orden uno del tipo

!           y'(x) = f(y(X),x)

! Este algoritmo es más preciso que el algoritmo de Euler al basarse en
! la derivada central en lugar de la derivada hacia delante.
 
        real function f(x, y)
            implicit none
            real :: x,y
            f = x*y
            return
        end  

        subroutine leap_frog(X, Y, N, dx)
            implicit none
            integer :: N, i
            real, dimension(N) :: X, Y
            real :: dx, f
            ! El segundo elemento inicial se calcula con la formula del
            ! método de euler y después los otros N-2 con leap-frog
            Y(2) = Y(1) + dx*f(X(1),Y(1))
            do i = 2, N-1
                Y(i+1) = Y(i-1) + 2*dx*f(X(i),Y(i))
            end do
        end subroutine

        program main

            implicit none
            integer :: i, N
            real :: xmin, xmax, dx, xo, yo
            real, allocatable :: X(:), Y(:)

            ! Definimos la malla de integración
            dx = 0.001
            xmin = 0.0
            xmax = 1.0
            N = (xmax - xmin)/dx + 1

            ! Condicion inicial
            xo = xmin
            yo = 2

            ! Definimos el tamaño del arreglo
            allocate(X(1:N), Y(1:N))
            do i = 0, N-1
                X(i+1) = xmin + i*dx
                Y(i+1) = 0
            end do
            Y(1) = yo

            ! Llamamos el método de Euler
            call leap_frog(X, Y, N, dx)

            ! write array to a file
            open(10, file='leap_frog.dat')
            do i=1,N
                write(10, '(f8.6,A4,f8.6)')X(i), "    " ,Y(i)
            end do
            close(10)

            ! Despejamos los arreglos de la memoria
            deallocate(X,Y)

        end program main

File: numerical_methods/edo_ivp/leapfrog_O1/leap_frog.py
# Programa elaborado por Jesús Eduardo Loera Casas
# Elaborado el día 30/01/23
 
import numpy as np
import matplotlib.pyplot as plt

"""
En este programa se implementa el método de Leap-Frog para resolver
ecuaciones diferenciales de la forma: 

        y'(x) = f(y(x), x)  con y(xo)=yo
"""   

"""
Función f(y(x), x)
"""
def funcion(y,x):
    return x*y

""""
Condiciones iniciales
"""
xo = 0
yo = 2

"""
Intervalo de integración
"""
xmin = xo
xmax = 1
N = 1000
X = np.linspace(xmin, xmax, N)

"""
Método de Euler
"""
def euler(X, xo, yo):
    Y = [yo]
    dx = X[1] - X[0]
    x = xo
    y = yo
    for i in range(1, len(X)):
        xi = x + dx
        yi = y + dx*funcion(y,x)
        Y.append(yi)
        x = xi
        y = yi
    return X, Y

"""
Método de Leap Frog
"""
def leapfrog(X, xo, yo):
    Y = [yo]
    dx = X[1] - X[0]
    Y.append(yo + dx*funcion(yo, xo))
    for i in range(1, len(X) - 1):
        y = Y[i-1] + 2*dx*funcion(X[i],Y[i])
        Y.append(y)
    return X, Y

Sol1 = euler(X, xo, yo)
Sol2 = leapfrog(X, xo, yo)
Y = 2*np.exp((X**2)/2)


plt.plot(Sol1[0], Sol1[1], label="Método de Euler", c= "blue")
plt.plot(Sol2[0], Sol2[1], label="Método de Leap Frog", c= "green")
plt.plot(X, Y, label ="Solución Analítica", c="orange")
plt.legend()
plt.grid()
plt.show()
